plan/logical_plan_builder.go

```diff
@@ -668,31 +668,86 @@ func (b *planBuilder) buildProjection4Union(u *LogicalUnionAll) {
 }
 
 func (b *planBuilder) buildUnion(union *ast.UnionStmt) LogicalPlan {
-	u := LogicalUnionAll{}.init(b.ctx)
-	u.children = make([]LogicalPlan, len(union.SelectList.Selects))
-	for i, sel := range union.SelectList.Selects {
-		u.children[i] = b.buildSelect(sel)
-		if b.err != nil {
-			return nil
-		}
-		if u.children[i].Schema().Len() != u.children[0].Schema().Len() {
-			b.err = errors.New("The used SELECT statements have a different number of columns")
-			return nil
-		}
+	distinctSelectPlans, allSelectPlans := b.resolveUnionSelectPlans(union.SelectList.Selects)
+	if b.err != nil {
+		return nil
 	}
 
-	b.buildProjection4Union(u)
-	var p LogicalPlan = u
-	if union.Distinct {
-		p = b.buildDistinct(u, u.Schema().Len())
+	unionDistinctPlan := b.buildSubUnion(distinctSelectPlans)
+	if b.err != nil {
+		return nil
 	}
+
+	if unionDistinctPlan != nil {
+		unionDistinctPlan = b.buildDistinct(unionDistinctPlan, unionDistinctPlan.Schema().Len())
+	}
+
+	if unionDistinctPlan != nil && allSelectPlans != nil && len(allSelectPlans) != 0 {
+		allSelectPlans = append(allSelectPlans, unionDistinctPlan)
+	}
+
+	unionAllPlan := b.buildSubUnion(allSelectPlans)
+	if b.err != nil {
+		return nil
+	}
+
+	var unionPlan LogicalPlan
+	if unionAllPlan != nil {
+		unionPlan = unionAllPlan
+	} else if unionDistinctPlan != nil {
+		unionPlan = unionDistinctPlan
+	} else {
```

Comment:
Will the else branch happen?
Can we this if-else to:
``` go
if unionAllPlan != nil {
	return unionAllPlan
} 
return unionDistinctPlan
```

Reply:
in first version, else will happen, but now we check `b.err` to fail fast return, so it never happen, I will remove that~